_getline.c

```diff
@@ -0,0 +1,106 @@
+#include "shell.h"
+
+/**
+ * _getline - read a line from a file descriptor and allocate memory for it.
+ * @fd: the file descriptor from which to read the line.
+ * @line: a pointer to a char pointer where the resulting line will be stored.
+ *
+ * This function reads a line from the specified file descriptor and allocates
+ * memory to store the line. The memory for the line should be freed by the
+ * caller when it is no longer needed.
+ *
+ * Return: Upon success, the function returns the number of characters read,
+ * excluding the null-terminator. On end of file or error, it returns -1.
+ */
+
+int _getline(char **line, const int fd)
+{
+	static char *rd = NULL;
+	char buffer[SIZE];
+	int pos;
+
+	pos = _strchr(rd, '\n');
+
+	if (pos >= 0)
+	{
+		process_line(&*line, &rd, pos);
+		return (pos);
+	}
+
+	while ((pos = _read(fd, &rd, buffer)) >= 0)
+	{
+		if (pos == 0)/*this means that we reach the end of file
+			      or the user pressed Ctrl+D*/
+		{
+			free(rd);
+			return(-2);
+		}
+		if (pos > 0 || pos < SIZE)
+			break;
+	}
+
+	if (pos == -1)
+	{
+		free(rd);
+		return (-1); /*reading Error*/
+	}
+
+	pos = _strchr(rd, '\n');
+
+	if (pos >= 0)/*pos == 0 here means that the newline charcter is found
+		       at the beginning of the buffer*/
+	{
+		process_line(&*line, &rd, pos);
+		return (pos);
+	}
+
+	*line = NULL;
+	free(rd);
+	return (pos); /*pos here will have a value less thhan 0*/
+}
+
+/**
+ * _read - read data from a file descriptor and append it to existing content.
+ * @fd: The file descriptor from which to read data.
+ * @rd: a char pointer representing existing content.
+ * @buff: a buffer to store the data read from the file descriptor.
+ *
+ * Return: Upon success, the function returns the number of bytes read.
+ * On error it returns -1.
+ */
+
+int _read(int fd, char **rd, char *buff)
+{
+	int readed = read(fd, buff, SIZE);
+
+	if (readed == -1)
+		return (-1); /*reading Error*/
+	if (readed > 0)
+	{
+		buff[readed] = '\0';
+		*rd = (*rd == NULL) ? _strdup(buff) : _strcat(*rd, buff);
+	}
+
+	return (readed);
+}
+
+/**
+ * process_line - Process a line from the existing content and
+ * update the pointers.
+ * @line: A pointer to a char pointer where the resulting line will be stored.
+ * @rd: A pointer to a char pointer representing existing content.
+ * @pos: The position of the newline character in the existing content.
+ */
+
+void process_line(char **line, char **rd, int pos)
+{
+	char *tmp;
+
+	*line = malloc((pos + 1) * sizeof(char));
+	_strncpy(*line, *rd, pos);
```

Comment:
I think if we have created functions similar to the C library at least we need to make sure they do similar things to use them in future projects, otherwise we give this function another name. 
please refer to the `man strncpy`.